doc(rfc): add feature gates design (#6199)
pingcap/tidb-operator: docs/rfcs/250515.md

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [CRD](#crd)
  - [Validation](#validation)
  - [Enable/Disable features](#enabledisable-features)
  - [Feature lifecycle](#feature-lifecycle)
    - [Alpha](#alpha)
  - [Test Plan](#test-plan)
  - [Feature Gate](#feature-gate)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Add features into group/instance annotations](#add-features-into-groupinstance-annotations)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a release*.

- [ ] (R) This design doc has been discussed and approved
- [ ] (R) Test plan is in place
  - [ ] (R) e2e tests in kind
- [ ] (R) Graduation criteria is in place if required
- [ ] (R) User-facing documentation has been created in [pingcap/docs-tidb-operator]

## Summary

TiDB Operator needs feature gates to control new features.

## Motivation

> It's important to avoid restarting TiDB clusters unexpectedly because it may influence the QoS of TiDB clusters. However, TiDB Operator have to restart pods if generated pod spec is changed. It's dangerous to develope and upgrade TiDB Operator.

Reviewer: ```suggestion
It's important to avoid restarting TiDB clusters unexpectedly because it may influence the QoS of TiDB clusters. However, TiDB Operator has to restart pods if the generated pod's spec is changed. It's dangerous to develop and upgrade TiDB Operator.
```

  - For TiDB, we have to indicate whether enable/disable a feature gate will restart TiDB pods.

## Design Details

### CRD

Add a new field feature gates into cluster.

```go
type Feature string

type ClusterSpec struct {
    // +listType=map
    // +listMapKey=name
    FeatureGates []FeatureGate `json:"featureGates,omitmepty"`
}

type FeatureGate struct {
    Name Feature `json:"name"`
}

type ClusterStatus struct {
    // +listType=map
    // +listMapKey=name
    FeatureGates []FeatureGateStatus `json:"featureGates,omitmepty"`
}

type FeatureGateStatus struct {
    Name Feature `json:"name"`
    Stage FeatureGateStage `json:"stage"`
    Restart bool `json:"restart,omitmepty"`
    // listType=set
    Components []string `json:"components,omitmepty"`
}

```

```yaml
apiVersion: core.pingcap.com/v1alpha1
kind: Cluster
spec:
  featureGates:
  - name: "ModifyVolumeAttribute"
status:
  # only show enabled features
  featureGates:
  - name: ModifyVolumeAttribute
    stage: ALPHA
  - name: StableFeature
    stage: STABLE
  - name: DisablePDReadinessProbe
    stage: ALPHA
    restart: true
    # Only change features of PD
    components:
    - PD
```

> Add a new annotation for both Group/Instance CRD

Reviewer: ```suggestion
Add a new field for both Group/Instance CRD
```
